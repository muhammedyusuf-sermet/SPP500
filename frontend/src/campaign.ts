export interface ICampaignState {
	Id?: string,
	Name: string,
	Description: string;
	// Monsters: Monster.IMonsterState[];
}
