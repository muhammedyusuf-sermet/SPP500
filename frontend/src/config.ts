export const API_URL = "https://http://3.18.65.138:3000";