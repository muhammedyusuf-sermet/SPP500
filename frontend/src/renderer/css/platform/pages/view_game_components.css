.view-game-components-container {
	text-align: center;
}
.page-title {
	padding: 5px 15px;
	display: inline-block;
}

.view-game-components-container .tab-root {
	flex-grow: 1;
}

.view-game-components-container .app-bar {
	position: relative;
}

.view-game-components-container .layout {
	width: auto;
	margin-left: 2%;
	margin-right: 2%;
}

.view-game-components-container .card-grid {
	padding: 8px 0;
}

.view-game-components-container .card {
	height: 100%;
	display: flex;
	flex-direction: column;
}

.view-game-components-container .card-media {
	padding-top: 56.25%; /* 16:9 */
}

.view-game-components-container .card-content {
	flex-grow: 1;
}
