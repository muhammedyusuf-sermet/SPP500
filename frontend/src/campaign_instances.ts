import * as Campaign from "./campaign";

// Dummy array of monsters
import EncounterInstances from './encounter_instances';

let encounterExample: Campaign.ICampaignState;
encounterExample = {
	Id: 0,
	Name: "Campaign",
	Summary: "Bloody hell...",
	Notes: "You gonna die!",
	Encounters: EncounterInstances['EncounterInstances'],
};

let CampaignInstances = new Array();

// Generate a dummy array of encounters with different IDs
for (var i = 0; i < 12; i++) {
	let encounter = {} as Campaign.ICampaignState;
	Object.assign(encounter, encounterExample);
	encounter['Id'] = i;
	CampaignInstances.push(encounter);
}

export default {
    CampaignInstances,
}