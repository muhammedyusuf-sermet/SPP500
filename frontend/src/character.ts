import { ICampaignData } from "./campaign";

export enum CharacterRace {
	Dragonborn = "Dragonborn",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Gnome = "Gnome",
	HalfElf = "HalfElf",
	Halfling = "Halfling",
	HalfOrc = "HalfOrc",
	Human = "Human",
	Tiefling = "Tiefling"
}

export enum CharacterClass {
	Barbarian = "Barbarian",
	Bard = "Bard",
	Cleric = "Cleric",
	Druid = "Druid",
	Fighter = "Fighter",
	Monk = "Monk",
	Paladin = "Paladin",
	Ranger = "Ranger",
	Rogue = "Rogue",
	Sorcerer = "Sorcerer",
	Warlock = "Warlock",
	Wizard = "Wizard"
}

export interface ICharacterData {
	Id?: number;
	Name: string;
	Level?: number;
	Race?: CharacterRace;
	Class?: CharacterClass;
	MaxHealth?: number;
	ArmorClass?: number;
	Notes?: string;

	Campaigns?: ICampaignData[];
}
