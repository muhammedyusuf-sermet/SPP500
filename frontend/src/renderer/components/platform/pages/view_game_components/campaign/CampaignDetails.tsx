import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import { stateWithoutErrors } from '../../../../../../utils/StateSelection';
import { ICampaignData } from '../../../../../../campaign';
import { IEncounterData } from '../../../../../../encounter';

export interface ICampaignDetailsProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Name?: string;
		Summary?: string;
		Notes?: string;
		Encounters?: IEncounterData[];
	}
}

export interface ICampaignDetailsState {
	// data
	Name?: string;
	Summary?: string;
	Notes?: string;
	Encounters?: string;
	// errors
	NameError?: string;
	SummaryError?: string;
	NotesError?: string;
	EncountersError?: string;
}

export class CampaignDetails extends React.Component<ICampaignDetailsProps, ICampaignDetailsState> {
	constructor(props: ICampaignDetailsProps) {
		super(props);
		this.state = {
			...props.initial,
			Encounters: props.initial.Encounters ? props.initial.Encounters.map((value)=>(value.Id)).join(',') : ''
		};
	}

	componentWillReceiveProps(nextProps: ICampaignDetailsProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial,
				Encounters: nextProps.initial.Encounters ? nextProps.initial.Encounters.map((value)=>(value.Id)).join(',') : ''
			});
	}

	handleStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = /^\s*$/.test(event.target.value) ? undefined : event.target.value;
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
		});
	}

	handleCampaignEncounterIdChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({
			Encounters: event.target.value
		})
	}

	render() {
		return (
			<Grid container spacing={8} >
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Name">Name</InputLabel>
						<Input
							id='Name'
							type='text'
							value={this.state.Name || ''}
							name='Name'
							onChange={this.handleStringChange}
							aria-describedby="Name-helper-text" />
						<FormHelperText error id="Name-helper-text">{this.state.NameError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Summary">Summary</InputLabel>
						<Input
							id='Summary'
							type='text'
							multiline
							rows={10}
							rowsMax={10}
							value={this.state.Summary || ''}
							name='Summary'
							onChange={this.handleStringChange}
							aria-describedby="Summary-helper-text" />
						<FormHelperText error id="Summary-helper-text">{this.state.SummaryError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Notes">Notes</InputLabel>
						<Input
							id='Notes'
							type='text'
							multiline
							rows={10}
							rowsMax={10}
							value={this.state.Notes || ''}
							name='Notes'
							onChange={this.handleStringChange}
							aria-describedby="Notes-helper-text" />
						<FormHelperText error id="Notes-helper-text">{this.state.NotesError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Encounters">Encounter Id's</InputLabel>
						<Input
							id='Encounters'
							type='text'
							multiline
							rows={10}
							rowsMax={10}
							value={this.state.Encounters || ''}
							name='Encounters'
							onChange={this.handleCampaignEncounterIdChange}
							aria-describedby="Encounters-helper-text" />
						<FormHelperText error id="Encounters-helper-text">{this.state.EncountersError}</FormHelperText>
					</FormControl>
				</Grid>
			</Grid>
		);
	}
}