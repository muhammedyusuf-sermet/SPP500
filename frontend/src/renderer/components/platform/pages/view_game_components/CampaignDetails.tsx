import * as React from 'react';
import * as request from 'request';
import { Input, Control, Label, Button, Tile, Subtitle, Modal, ModalContent, ModalBackground, Box } from 'bloomer';
import { API_URL } from "../../../../../config"
import { CookieManager } from "../../../../../cookie";
import * as EncounterInterface from '../../../../../encounter';
import * as CampaignInterface from '../../../../../campaign';

import 'bulma/css/bulma.css';

interface ICampaignDeleteResponse {
	status: number,
	messages: string[]
}

export interface ICampaignDetailsState {
	campaign: CampaignInterface.ICampaignState,
	modal: {
		open: boolean,
		message: string
	}
}

export class CampaignDetails extends React.Component<any, ICampaignDetailsState> {
	constructor(props: any) {
		super(props);
		this.state = {
			campaign: this.props.campaign,
			modal: {
				open: false,
				message: ""
			}
		}
		this.preprocessMissingCampaignData();
	}

	isEmptyObject(obj: CampaignInterface.ICampaignState) {
		for(var key in obj) {
			if(obj.hasOwnProperty(key))
				return false;
		}
		return true;
	}

	// For now, we are not getting the encounters of an campaign from the backend.
	// Once the backend is fixed, we can get rid of this function.
	preprocessMissingCampaignData() {
		let campaign = this.state.campaign;
		if(campaign.Encounters == null){
			campaign.Encounters = [] as EncounterInterface.IEncounterState[];
		}
	}

	openModal = (messageText: string) => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: true, message: messageText }});
	};

	closeModal = () => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: false }});
		this.props.resetParentState();
	};

	deleteCampaign = (event: React.FormEvent) => {
		event.preventDefault();
		let campaign = this.state.campaign;
		var options = { method: 'POST',
			url: API_URL + '/campaign/delete',
			timeout: 2000,
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json' ,
				'Authorization': CookieManager.UserToken('session_token')
			},
			body:
			{
				Id: campaign.Id,
			},
			json: true
		};
		request(options, (error: string, response: any, body: ICampaignDeleteResponse) => {
			if (error) {
				this.openModal("There has been a server error when deleting the campaign. Please try again later.");
			} else {
				let { status, messages } = body
				if (status == 201){
					this.openModal("The campaign is successfully deleted! Click anywhere to be redirected to the catalogue.");
				}
				else{
					let message = "There has been an error deleting the campaign. "
					if (messages) message = messages.join(' ');
					this.openModal(message);
				}
			}
		});
	}

	isExists(attribute: string | number | undefined) {
		if(attribute == undefined){
			return "N/A";
		}
		else{
			return attribute;
		}
	}

	render() {
		if(!this.isEmptyObject(this.state.campaign)) {
			let campaign = this.state.campaign;
			return (
				<div>
					<form>
						<h1 className="page-title">Campaign Details</h1>
							<Control>
								<Label>Campaign Name</Label>
								<Input  value={campaign.Name} id="name" label="Campaign Name" readOnly/>
							</Control>
							<Control>
								<Label>Campaign Summary</Label>
								<Input  value={campaign.Summary} id="description" label="Campaign Summary" readOnly/>
							</Control>
							<Control>
								<Label>Campaign Notes</Label>
								<Input  value={campaign.Notes} id="description" label="Campaign Notes" readOnly/>
							</Control>
							<Tile className="box" isVertical>
								<Subtitle>Encounters</Subtitle>
								{/* {campaign.Encounters.map(encounter => (
									<Control key={encounter.Id}>
										<Label>{this.isExists(encounter.Name)} {this.isExists(encounter.Id)}</Label>
									</Control>
								))} */}
							</Tile>
						<Button className="button" id="deleteCampaignButton" onClick = {this.deleteCampaign}> Delete Campaign </Button>
						<Button className="button" type="submit" disabled> Edit Campaign </Button>
						<Button className="button" onClick = {this.props.resetParentState}> Return to Game Components </Button>
					</form>
					<Modal id='campaignDeletionModal' isActive={this.state.modal.open}>
						<ModalBackground id='modalBackground' onClick={()=>{
							this.closeModal();
						}}/>
						<ModalContent>
							<Box>
								<span id="ModalMessage">{this.state.modal.message}</span>
							</Box>
						</ModalContent>
					</Modal>
				</div>
			);
		}
		else{
			return (<div id="errorMessageNoCampaign">No campaign is selected.</div>);
		}
	}
}