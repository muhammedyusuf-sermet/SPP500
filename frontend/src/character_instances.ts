import { CharacterRace, CharacterClass, ICharacterData } from "./character";


let characterExample: ICharacterData;
characterExample = {
	Id: 1,
	Name: "Character",
	Level: 3,
	Race: CharacterRace.Gnome,
	Class: CharacterClass.Paladin,
	MaxHealth: 23,
	ArmorClass: 12,
	Notes: 'aggressive',

	Campaigns: []
};

export const CharacterInstances: ICharacterData[] = new Array();

// Generate a dummy array of characters with different IDs
for (var i = 0; i < 12; i++) {
	let character = {} as ICharacterData;
	Object.assign(character, characterExample);
	character.Id = i + 1;
	CharacterInstances.push(character);
}