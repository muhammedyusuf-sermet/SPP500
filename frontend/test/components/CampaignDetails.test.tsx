import * as React from "react"
import * as nock from 'nock';
import { shallow, ShallowWrapper } from 'enzyme';
import {CampaignDetails, ICampaignDetailsState} from '../../src/renderer/components/platform/pages/view_game_components/CampaignDetails';
import * as Campaign from '../../src/campaign';
// import * as Monster from "../../src/monster";
import {API_URL} from '../../src/config'
import EncounterInstances from "../../src/encounter_instances";

jest.mock('../../src/cookie');

describe('Test the Campaign View Details Per Campaign', () => {
	let campaignDetailsInstance: ShallowWrapper<any, ICampaignDetailsState, CampaignDetails>;
	let campaignDetailsInstanceFaulty: ShallowWrapper<any, ICampaignDetailsState, CampaignDetails>;

	beforeEach(() => {
		let campaignExample: Campaign.ICampaignState;
		campaignExample = {
			Id: "1",
			Name: "Campaign",
			Notes: "Notes",
			Summary: "This is a fierce campaign...",
			Encounters: EncounterInstances['EncounterInstances'],
		};
		campaignDetailsInstance = shallow(<CampaignDetails campaign={campaignExample} resetParentState={() => {return null;}}/>);
	})

	it('renders without crashing', () => {
		expect(campaignDetailsInstance).toBeDefined();
	});

	it('renders correctly when the page is loaded', () => {
		expect(campaignDetailsInstance).toMatchSnapshot();
	});

	it('should return true if the ICampaignState object is empty', () => {
		let selectedCampaign: Campaign.ICampaignState;
		selectedCampaign = {} as Campaign.ICampaignState;
		let isEmpty = campaignDetailsInstance.instance().isEmptyObject(selectedCampaign);

		expect(isEmpty).toEqual(true);
	});

	it('successfully deletes the campaign', () => {
		var deleteButton = campaignDetailsInstance.find('#deleteCampaignButton');

		nock(API_URL)
			.post('/campaign/delete', {
					Id: "1",
				})
			.reply(201, {
				body: [{ status: 201, messages: 'success' }],
			});

		deleteButton.simulate('click', {preventDefault: () => {}});
	});

	describe('should show and hide modal', () => {
		it('show modal', () => {
			campaignDetailsInstance.instance().closeModal();
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(false);
			campaignDetailsInstance.instance().openModal('TestMessage');
			expect(campaignDetailsInstance.find('#ModalMessage').text()).toEqual('TestMessage');
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(true);
		})

		it('close modal', () => {
			campaignDetailsInstance.instance().openModal('TestMessage');
			expect(campaignDetailsInstance.find('#ModalMessage').text()).toEqual('TestMessage');
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(true);
			campaignDetailsInstance.instance().closeModal();
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(false);
		})

		it('close modal by click', () => {
			campaignDetailsInstance.instance().openModal('TestMessage');
			expect(campaignDetailsInstance.find('#ModalMessage').text().length).toBeGreaterThan(0);
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(true);
			let background = campaignDetailsInstance.find('#modalBackground')
			background.simulate('click');
			expect(campaignDetailsInstance.find('#campaignDeletionModal').prop('isActive')).toEqual(false);
		})
	});

	describe('Sad Paths', () => {
		beforeEach(() => {
			campaignDetailsInstanceFaulty = shallow(<CampaignDetails campaign={{}} />);
		})

		it('displays error message when no campaign is selected', () => {
			let selectedCampaign: Campaign.ICampaignState;
			selectedCampaign = {} as Campaign.ICampaignState;
			campaignDetailsInstanceFaulty.instance().setState({ campaign: selectedCampaign});
			expect(campaignDetailsInstanceFaulty.find('#errorMessageNoCampaign')).toExist();
		});
	});
});
