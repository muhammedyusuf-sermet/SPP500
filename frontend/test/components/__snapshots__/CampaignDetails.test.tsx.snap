// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Campaign View Details Per Campaign renders correctly when the page is loaded 1`] = `
<div>
  <form>
    <h1
      className="page-title"
    >
      Campaign Details
    </h1>
    <withHelpersModifiers(Control)>
      <withHelpersModifiers(Label)>
        Campaign Name
      </withHelpersModifiers(Label)>
      <withHelpersModifiers(Input)
        id="name"
        label="Campaign Name"
        readOnly={true}
        value="Campaign"
      />
    </withHelpersModifiers(Control)>
    <withHelpersModifiers(Control)>
      <withHelpersModifiers(Label)>
        Campaign Summary
      </withHelpersModifiers(Label)>
      <withHelpersModifiers(Input)
        id="description"
        label="Campaign Summary"
        readOnly={true}
        value="This is a fierce campaign..."
      />
    </withHelpersModifiers(Control)>
    <withHelpersModifiers(Control)>
      <withHelpersModifiers(Label)>
        Campaign Notes
      </withHelpersModifiers(Label)>
      <withHelpersModifiers(Input)
        id="description"
        label="Campaign Notes"
        readOnly={true}
        value="Notes"
      />
    </withHelpersModifiers(Control)>
    <withHelpersModifiers(Tile)
      className="box"
      isVertical={true}
    >
      <withHelpersModifiers(Subtitle)>
        Encounters
      </withHelpersModifiers(Subtitle)>
      <withHelpersModifiers(Control)
        key="0"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          0
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="1"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          1
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="2"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          2
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="3"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          3
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="4"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          4
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="5"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          5
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="6"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          6
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="7"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          7
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="8"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          8
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="9"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          9
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="10"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          10
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        key="11"
      >
        <withHelpersModifiers(Label)>
          Encounter
           
          11
        </withHelpersModifiers(Label)>
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Tile)>
    <withHelpersModifiers(Button)
      className="button"
      id="deleteCampaignButton"
      onClick={[Function]}
    >
       Delete Campaign 
    </withHelpersModifiers(Button)>
    <withHelpersModifiers(Button)
      className="button"
      disabled={true}
      type="submit"
    >
       Edit Campaign 
    </withHelpersModifiers(Button)>
    <withHelpersModifiers(Button)
      className="button"
      onClick={[Function]}
    >
       Return to Game Components 
    </withHelpersModifiers(Button)>
  </form>
  <withHelpersModifiers(Modal)
    id="campaignDeletionModal"
    isActive={false}
  >
    <withHelpersModifiers(ModalBackground)
      id="modalBackground"
      onClick={[Function]}
    />
    <withHelpersModifiers(ModalContent)>
      <withHelpersModifiers(Box)>
        <span
          id="ModalMessage"
        />
      </withHelpersModifiers(Box)>
    </withHelpersModifiers(ModalContent)>
  </withHelpersModifiers(Modal)>
</div>
`;
